<p style="text-align:center">
    <a href="https://skills.network/?utm_medium=Exinfluencer&utm_source=Exinfluencer&utm_content=000026UJ&utm_term=10006555&utm_id=NA-SkillsNetwork-Channel-SkillsNetworkCoursesIBMDeveloperSkillsNetworkPY0220ENSkillsNetwork900-2022-01-01" target="_blank">
    <img src="https://cf-courses-data.s3.us.cloud-object-storage.appdomain.cloud/assets/logos/SN_web_lightmode.png" width="200" alt="Skills Network Logo">
    </a>
</p>


<h1>Extracting and Visualizing Stock Data</h1>
<h2>Description</h2>


Extracting essential data from a dataset and displaying it is a necessary part of data science; therefore individuals can make correct decisions based on the data. In this assignment, you will extract some stock data, you will then display this data in a graph.


<h2>Table of Contents</h2>
<div class="alert alert-block alert-info" style="margin-top: 20px">
    <ul>
        <li>Define a Function that Makes a Graph</li>
        <li>Question 1: Use yfinance to Extract Stock Data</li>
        <li>Question 2: Use Webscraping to Extract Tesla Revenue Data</li>
        <li>Question 3: Use yfinance to Extract Stock Data</li>
        <li>Question 4: Use Webscraping to Extract GME Revenue Data</li>
        <li>Question 5: Plot Tesla Stock Graph</li>
        <li>Question 6: Plot GameStop Stock Graph</li>
    </ul>
<p>
    Estimated Time Needed: <strong>30 min</strong></p>
</div>

<hr>


***Note***:- If you are working Locally using anaconda, please uncomment the following code and execute it.


In [1]:
#!pip install yfinance==0.2.38
#!pip install pandas==2.2.2
#!pip install nbformat

In [2]:
!pip install yfinance==0.1.67
!mamba install bs4==4.10.0 -y
!pip install nbformat==4.2.0


                  __    __    __    __
                 /  \  /  \  /  \  /  \
                /    \/    \/    \/    \
███████████████/  /██/  /██/  /██/  /████████████████████████
              /  / \   / \   / \   / \  \____
             /  /   \_/   \_/   \_/   \    o \__,
            / _/                       \_____/  `
            |/
        ███╗   ███╗ █████╗ ███╗   ███╗██████╗  █████╗
        ████╗ ████║██╔══██╗████╗ ████║██╔══██╗██╔══██╗
        ██╔████╔██║███████║██╔████╔██║██████╔╝███████║
        ██║╚██╔╝██║██╔══██║██║╚██╔╝██║██╔══██╗██╔══██║
        ██║ ╚═╝ ██║██║  ██║██║ ╚═╝ ██║██████╔╝██║  ██║
        ╚═╝     ╚═╝╚═╝  ╚═╝╚═╝     ╚═╝╚═════╝ ╚═╝  ╚═╝

        mamba (1.4.2) supported by @QuantStack

        GitHub:  https://github.com/mamba-org/mamba
        Twitter: https://twitter.com/QuantStack

█████████████████████████████████████████████████████████████


Looking for: ['bs4==4.10.0']

[+] 0.0s
[+] 0.1s
pkgs/main/linux-64 ━━━━━╸━━━━━━━━━━━━━━━╸━━━   0.0 B /  ??.?MB @ 

In [3]:
import yfinance as yf
import pandas as pd
import requests
from bs4 import BeautifulSoup
import plotly.graph_objects as go
from plotly.subplots import make_subplots

In Python, you can ignore warnings using the warnings module. You can use the filterwarnings function to filter or ignore specific warning messages or categories.


In [4]:
import warnings
# Ignore all warnings
warnings.filterwarnings("ignore", category=FutureWarning)

## Define Graphing Function


In this section, we define the function `make_graph`. **You don't have to know how the function works, you should only care about the inputs. It takes a dataframe with stock data (dataframe must contain Date and Close columns), a dataframe with revenue data (dataframe must contain Date and Revenue columns), and the name of the stock.**


In [5]:
def make_graph(stock_data, revenue_data, stock):
    fig = make_subplots(rows=2, cols=1, shared_xaxes=True, subplot_titles=("Historical Share Price", "Historical Revenue"), vertical_spacing = .3)
    stock_data_specific = stock_data[stock_data.Date <= '2021--06-14']
    revenue_data_specific = revenue_data[revenue_data.Date <= '2021-04-30']
    fig.add_trace(go.Scatter(x=pd.to_datetime(stock_data_specific.Date, infer_datetime_format=True), y=stock_data_specific.Close.astype("float"), name="Share Price"), row=1, col=1)
    fig.add_trace(go.Scatter(x=pd.to_datetime(revenue_data_specific.Date, infer_datetime_format=True), y=revenue_data_specific.Revenue.astype("float"), name="Revenue"), row=2, col=1)
    fig.update_xaxes(title_text="Date", row=1, col=1)
    fig.update_xaxes(title_text="Date", row=2, col=1)
    fig.update_yaxes(title_text="Price ($US)", row=1, col=1)
    fig.update_yaxes(title_text="Revenue ($US Millions)", row=2, col=1)
    fig.update_layout(showlegend=False,
    height=900,
    title=stock,
    xaxis_rangeslider_visible=True)
    fig.show()

Use the make_graph function that we’ve already defined. You’ll need to invoke it in questions 5 and 6 to display the graphs and create the dashboard. 
> **Note: You don’t need to redefine the function for plotting graphs anywhere else in this notebook; just use the existing function.**


## Question 1: Use yfinance to Extract Stock Data


Using the `Ticker` function enter the ticker symbol of the stock we want to extract data on to create a ticker object. The stock is Tesla and its ticker symbol is `TSLA`.


In [6]:
import yfinance as yf

# Create a ticker object for Tesla
tesla = yf.Ticker("TSLA")

# Get historical market data
hist = tesla.history(period="1mo")

# Show the historical market data
print(hist)

# Get stock information
info = tesla.info

# Show stock information
print(info)


                  Open        High         Low       Close     Volume  \
Date                                                                    
2024-05-28  176.399994  178.250000  173.160004  176.750000   59736600   
2024-05-29  174.190002  178.149994  173.929993  176.190002   54782600   
2024-05-30  178.580002  182.669998  175.380005  178.789993   77784800   
2024-05-31  178.500000  180.320007  173.820007  178.080002   67314600   
2024-06-03  178.130005  182.639999  174.490005  176.289993   68568900   
2024-06-04  174.779999  177.759995  174.000000  174.770004   60056300   
2024-06-05  175.350006  176.149994  172.130005  175.000000   57953800   
2024-06-06  174.600006  179.729996  172.729996  177.940002   69887000   
2024-06-07  176.130005  179.350006  175.580002  177.479996   56244900   
2024-06-10  176.059998  178.570007  173.169998  173.789993   50869700   
2024-06-11  173.919998  174.750000  167.410004  170.660004   64761900   
2024-06-12  171.119995  180.550003  169.800003  177

Using the ticker object and the function `history` extract stock information and save it in a dataframe named `tesla_data`. Set the `period` parameter to ` "max" ` so we get information for the maximum amount of time.


In [7]:
import yfinance as yf
import pandas as pd

# Create a ticker object for Tesla
tesla = yf.Ticker("TSLA")

# Get historical market data for the maximum available period
tesla_data = tesla.history(period="max")

# Display the first few rows of the dataframe
print(tesla_data.head())


                Open      High       Low     Close     Volume  Dividends  \
Date                                                                       
2010-06-29  1.266667  1.666667  1.169333  1.592667  281494500          0   
2010-06-30  1.719333  2.028000  1.553333  1.588667  257806500          0   
2010-07-01  1.666667  1.728000  1.351333  1.464000  123282000          0   
2010-07-02  1.533333  1.540000  1.247333  1.280000   77097000          0   
2010-07-06  1.333333  1.333333  1.055333  1.074000  103003500          0   

            Stock Splits  
Date                      
2010-06-29           0.0  
2010-06-30           0.0  
2010-07-01           0.0  
2010-07-02           0.0  
2010-07-06           0.0  


**Reset the index** using the `reset_index(inplace=True)` function on the tesla_data DataFrame and display the first five rows of the `tesla_data` dataframe using the `head` function. Take a screenshot of the results and code from the beginning of Question 1 to the results below.


In [8]:
import yfinance as yf
import pandas as pd

# Create a ticker object for Tesla
tesla = yf.Ticker("TSLA")

# Get historical market data for the maximum available period
tesla_data = tesla.history(period="max")

# Reset the index of the dataframe
tesla_data.reset_index(inplace=True)

# Display the first five rows of the dataframe
print(tesla_data.head())


        Date      Open      High       Low     Close     Volume  Dividends  \
0 2010-06-29  1.266667  1.666667  1.169333  1.592667  281494500          0   
1 2010-06-30  1.719333  2.028000  1.553333  1.588667  257806500          0   
2 2010-07-01  1.666667  1.728000  1.351333  1.464000  123282000          0   
3 2010-07-02  1.533333  1.540000  1.247333  1.280000   77097000          0   
4 2010-07-06  1.333333  1.333333  1.055333  1.074000  103003500          0   

   Stock Splits  
0           0.0  
1           0.0  
2           0.0  
3           0.0  
4           0.0  


## Question 2: Use Webscraping to Extract Tesla Revenue Data


Use the `requests` library to download the webpage https://cf-courses-data.s3.us.cloud-object-storage.appdomain.cloud/IBMDeveloperSkillsNetwork-PY0220EN-SkillsNetwork/labs/project/revenue.htm Save the text of the response as a variable named `html_data`.


In [9]:
import requests

# URL of the webpage to download
url = "https://cf-courses-data.s3.us.cloud-object-storage.appdomain.cloud/IBMDeveloperSkillsNetwork-PY0220EN-SkillsNetwork/labs/project/revenue.htm"

# Send a GET request to download the webpage
response = requests.get(url)

# Save the text of the response as a variable named html_data
html_data = response.text

# Display the first 500 characters of the HTML data
print(html_data[:500])



<!DOCTYPE html>
<!--[if lt IE 7]>      <html class="no-js lt-ie9 lt-ie8 lt-ie7"> <![endif]-->
<!--[if IE 7]>         <html class="no-js lt-ie9 lt-ie8"> <![endif]-->
<!--[if IE 8]>         <html class="no-js lt-ie9"> <![endif]-->
<!--[if gt IE 8]><!--> <html class="no-js"> <!--<![endif]-->
    <head>
        <meta charset="utf-8">
        <meta http-equiv="X-UA-Compatible" content="IE=edge,chrome=1">
		<link rel="canonical" href="https://www.macrotrends.net/stocks/charts/TSLA/tesla/revenue" />
	


Parse the html data using `beautiful_soup` using parser i.e `html5lib` or `html.parser`.


In [10]:
import requests
from bs4 import BeautifulSoup

# URL of the webpage to download
url = "https://cf-courses-data.s3.us.cloud-object-storage.appdomain.cloud/IBMDeveloperSkillsNetwork-PY0220EN-SkillsNetwork/labs/project/revenue.htm"

# Send a GET request to download the webpage
response = requests.get(url)

# Save the text of the response as a variable named html_data
html_data = response.text

# Parse the HTML data using BeautifulSoup and the html.parser
soup = BeautifulSoup(html_data, 'html.parser')

# Print the parsed content
print(soup.prettify()[:1000])  # Displaying the first 1000 characters of the parsed content


<!DOCTYPE html>
<!--[if lt IE 7]>      <html class="no-js lt-ie9 lt-ie8 lt-ie7"> <![endif]-->
<!--[if IE 7]>         <html class="no-js lt-ie9 lt-ie8"> <![endif]-->
<!--[if IE 8]>         <html class="no-js lt-ie9"> <![endif]-->
<!--[if gt IE 8]><!-->
<html class="no-js">
 <!--<![endif]-->
 <head>
  <meta charset="utf-8"/>
  <meta content="IE=edge,chrome=1" http-equiv="X-UA-Compatible"/>
  <link href="https://www.macrotrends.net/stocks/charts/TSLA/tesla/revenue" rel="canonical"/>
  <title>
   Tesla Revenue 2010-2022 | TSLA | MacroTrends
  </title>
  <meta content="Tesla annual/quarterly revenue history and growth rate from 2010 to 2022. Revenue can be defined as the amount of money a company receives from its customers in exchange for the sales of goods or services.  Revenue is the top line item on an income statement from which all costs and expenses are subtracted to arrive at net income.    
				
				&lt;ul style='margin-top:10px;'&gt;
				&lt;li&gt;Tesla revenue for the quarter end

Using `BeautifulSoup` or the `read_html` function extract the table with `Tesla Revenue` and store it into a dataframe named `tesla_revenue`. The dataframe should have columns `Date` and `Revenue`.


<details><summary>Step-by-step instructions</summary>

```

Here are the step-by-step instructions:

1. Create an Empty DataFrame
2. Find the Relevant Table
3. Check for the Tesla Quarterly Revenue Table
4. Iterate Through Rows in the Table Body
5. Extract Data from Columns
6. Append Data to the DataFrame

```
</details>


<details><summary>Click here if you need help locating the table</summary>

```
    
Below is the code to isolate the table, you will now need to loop through the rows and columns like in the previous lab
    
soup.find_all("tbody")[1]
    
If you want to use the read_html function the table is located at index 1

We are focusing on quarterly revenue in the lab.
```

</details>


In [11]:
import requests
from bs4 import BeautifulSoup
import pandas as pd

# URL of the webpage to download
url = "https://cf-courses-data.s3.us.cloud-object-storage.appdomain.cloud/IBMDeveloperSkillsNetwork-PY0220EN-SkillsNetwork/labs/project/revenue.htm"

# Send a GET request to download the webpage
response = requests.get(url)

# Save the text of the response as a variable named html_data
html_data = response.text

# Parse the HTML data using BeautifulSoup
soup = BeautifulSoup(html_data, 'html.parser')

# Find the table containing Tesla Revenue
table = soup.find_all('table')[1]  # Assuming the second table contains the revenue data

# Read the table into a dataframe
tesla_revenue = pd.read_html(str(table))[0]

# Rename the columns to 'Date' and 'Revenue'
tesla_revenue.columns = ['Date', 'Revenue']

# Display the first few rows of the dataframe
print(tesla_revenue.head())


         Date  Revenue
0  2022-09-30  $21,454
1  2022-06-30  $16,934
2  2022-03-31  $18,756
3  2021-12-31  $17,719
4  2021-09-30  $13,757


Execute the following line to remove the comma and dollar sign from the `Revenue` column. 


In [12]:
tesla_revenue["Revenue"] = tesla_revenue['Revenue'].str.replace(',|\$',"")

Execute the following lines to remove an null or empty strings in the Revenue column.


In [13]:
tesla_revenue.dropna(inplace=True)

tesla_revenue = tesla_revenue[tesla_revenue['Revenue'] != ""]

Display the last 5 row of the `tesla_revenue` dataframe using the `tail` function. Take a screenshot of the results.


In [14]:
import requests
from bs4 import BeautifulSoup
import pandas as pd

# URL of the webpage to download
url = "https://cf-courses-data.s3.us.cloud-object-storage.appdomain.cloud/IBMDeveloperSkillsNetwork-PY0220EN-SkillsNetwork/labs/project/revenue.htm"

# Send a GET request to download the webpage
response = requests.get(url)

# Save the text of the response as a variable named html_data
html_data = response.text

# Parse the HTML data using BeautifulSoup
soup = BeautifulSoup(html_data, 'html.parser')

# Find the table containing Tesla Revenue
table = soup.find_all('table')[1]  # Assuming the second table contains the revenue data

# Read the table into a dataframe
tesla_revenue = pd.read_html(str(table))[0]

# Rename the columns to 'Date' and 'Revenue'
tesla_revenue.columns = ['Date', 'Revenue']

# Display the last five rows of the dataframe
print(tesla_revenue.tail())


          Date Revenue
49  2010-06-30     $28
50  2010-03-31     $21
51  2009-12-31     NaN
52  2009-09-30     $46
53  2009-06-30     $27


## Question 3: Use yfinance to Extract Stock Data


Using the `Ticker` function enter the ticker symbol of the stock we want to extract data on to create a ticker object. The stock is GameStop and its ticker symbol is `GME`.


In [15]:
import yfinance as yf

# Create a ticker object for GameStop
gamestop = yf.Ticker("GME")

# Print basic information about GameStop
print(gamestop.info)


{'regularMarketPrice': None, 'preMarketPrice': None, 'logo_url': ''}


Using the ticker object and the function `history` extract stock information and save it in a dataframe named `gme_data`. Set the `period` parameter to ` "max" ` so we get information for the maximum amount of time.


In [16]:
import yfinance as yf

# Create a ticker object for GameStop
gamestop = yf.Ticker("GME")

# Get historical market data for the maximum available period
gme_data = gamestop.history(period="max")

# Display the first few rows of the dataframe
print(gme_data.head())


                Open      High       Low     Close    Volume  Dividends  \
Date                                                                      
2002-02-13  1.620128  1.693350  1.603296  1.691666  76216000        0.0   
2002-02-14  1.712707  1.716074  1.670626  1.683251  11021600        0.0   
2002-02-15  1.683251  1.687459  1.658002  1.674834   8389600        0.0   
2002-02-19  1.666418  1.666418  1.578047  1.607504   7410400        0.0   
2002-02-20  1.615920  1.662210  1.603296  1.662210   6892800        0.0   

            Stock Splits  
Date                      
2002-02-13           0.0  
2002-02-14           0.0  
2002-02-15           0.0  
2002-02-19           0.0  
2002-02-20           0.0  


**Reset the index** using the `reset_index(inplace=True)` function on the gme_data DataFrame and display the first five rows of the `gme_data` dataframe using the `head` function. Take a screenshot of the results and code from the beginning of Question 3 to the results below.


In [17]:
import yfinance as yf

# Create a ticker object for GameStop
gamestop = yf.Ticker("GME")

# Get historical market data for the maximum available period
gme_data = gamestop.history(period="max")

# Reset the index of the dataframe
gme_data.reset_index(inplace=True)

# Display the first five rows of the dataframe
print(gme_data.head())


        Date      Open      High       Low     Close    Volume  Dividends  \
0 2002-02-13  1.620128  1.693350  1.603296  1.691666  76216000        0.0   
1 2002-02-14  1.712707  1.716074  1.670626  1.683251  11021600        0.0   
2 2002-02-15  1.683250  1.687458  1.658001  1.674834   8389600        0.0   
3 2002-02-19  1.666418  1.666418  1.578047  1.607504   7410400        0.0   
4 2002-02-20  1.615920  1.662210  1.603296  1.662210   6892800        0.0   

   Stock Splits  
0           0.0  
1           0.0  
2           0.0  
3           0.0  
4           0.0  


## Question 4: Use Webscraping to Extract GME Revenue Data


Use the `requests` library to download the webpage https://cf-courses-data.s3.us.cloud-object-storage.appdomain.cloud/IBMDeveloperSkillsNetwork-PY0220EN-SkillsNetwork/labs/project/stock.html. Save the text of the response as a variable named `html_data_2`.


In [19]:
import requests

# URL of the webpage
url = "https://cf-courses-data.s3.us.cloud-object-storage.appdomain.cloud/IBMDeveloperSkillsNetwork-PY0220EN-SkillsNetwork/labs/project/stock.html"

# Send a GET request to the URL
response = requests.get(url)

# Check if the request was successful (status code 200)
if response.status_code == 200:
    # Save the HTML content as a variable named html_data_2
    html_data_2 = response.text
    print("Webpage downloaded successfully.")
else:
    print(f"Failed to download webpage. Status code: {response.status_code}")


Webpage downloaded successfully.


Parse the html data using `beautiful_soup` using parser i.e `html5lib` or `html.parser`.


In [20]:
import requests
from bs4 import BeautifulSoup

# URL of the webpage
url = "https://cf-courses-data.s3.us.cloud-object-storage.appdomain.cloud/IBMDeveloperSkillsNetwork-PY0220EN-SkillsNetwork/labs/project/stock.html"

# Send a GET request to the URL
response = requests.get(url)

# Check if the request was successful (status code 200)
if response.status_code == 200:
    # Parse the HTML using BeautifulSoup with html.parser
    soup = BeautifulSoup(response.text, 'html.parser')
    
    # Now you can work with the parsed HTML
    # For example, print the title of the webpage
    print("Title of the webpage:", soup.title.string)
else:
    print(f"Failed to download webpage. Status code: {response.status_code}")


Title of the webpage: GameStop Revenue 2006-2020 | GME | MacroTrends


Using `BeautifulSoup` or the `read_html` function extract the table with `GameStop Revenue` and store it into a dataframe named `gme_revenue`. The dataframe should have columns `Date` and `Revenue`. Make sure the comma and dollar sign is removed from the `Revenue` column.


> **Note: Use the method similar to what you did in question 2.**  


<details><summary>Click here if you need help locating the table</summary>

```
    
Below is the code to isolate the table, you will now need to loop through the rows and columns like in the previous lab
    
soup.find_all("tbody")[1]
    
If you want to use the read_html function the table is located at index 1


```

</details>


In [35]:
import requests
from bs4 import BeautifulSoup
import pandas as pd

# Define the URL of the webpage containing GameStop revenue data
url = "https://cf-courses-data.s3.us.cloud-object-storage.appdomain.cloud/IBMDeveloperSkillsNetwork-PY0220EN-SkillsNetwork/labs/project/stock.html"

# Function to retrieve and parse GameStop Revenue Data using BeautifulSoup
def retrieve_gme_revenue_data(url):
    response = requests.get(url)
    if response.status_code == 200:
        soup = BeautifulSoup(response.text, 'html.parser')
        # Find the table containing GameStop Revenue Data
        table = soup.find('table', {'id': 'gamestop-revenue'})  # Assuming the table has an id attribute 'gamestop-revenue'
        
        if table is None:
            print("Could not find GameStop Revenue Data table.")
            return None
        
        # Extract table rows
        rows = table.find_all('tr')
        
        # Extract table headers
        headers = [header.text.strip() for header in rows[0].find_all('th')]
        
        # Extract table data and create DataFrame
        data = []
        for row in rows[1:]:
            data.append([cell.text.strip() for cell in row.find_all('td')])
        
        # Create DataFrame
        gme_revenue = pd.DataFrame(data, columns=headers)
        
        # Clean up the Revenue column (remove commas and dollar signs)
        gme_revenue['Revenue'] = gme_revenue['Revenue'].str.replace(',', '').str.replace('$', '')
        
        # Convert Revenue column to numeric
        gme_revenue['Revenue'] = pd.to_numeric(gme_revenue['Revenue'], errors='coerce')
        
        # Convert Date column to datetime
        gme_revenue['Date'] = pd.to_datetime(gme_revenue['Date'])
        
        # Select only Date and Revenue columns
        gme_revenue = gme_revenue[['Date', 'Revenue']]
        
        return gme_revenue
    else:
        print(f"Failed to download webpage. Status code: {response.status_code}")
        return None

# Retrieve GameStop revenue data and store it in gme_revenue DataFrame
gme_revenue = retrieve_gme_revenue_data(url)

# Display the DataFrame
if gme_revenue is not None:
    print("GameStop Revenue DataFrame:")
    print(gme_revenue.head())
else:
    print("Failed to retrieve GameStop Revenue Data.")


Could not find GameStop Revenue Data table.
Failed to retrieve GameStop Revenue Data.


Display the last five rows of the `gme_revenue` dataframe using the `tail` function. Take a screenshot of the results.


In [23]:
import requests
from bs4 import BeautifulSoup
import pandas as pd

# URL of the webpage
url = "https://cf-courses-data.s3.us.cloud-object-storage.appdomain.cloud/IBMDeveloperSkillsNetwork-PY0220EN-SkillsNetwork/labs/project/stock.html"

# Send a GET request to the URL
response = requests.get(url)

# Check if the request was successful (status code 200)
if response.status_code == 200:
    # Parse the HTML using BeautifulSoup with html.parser
    soup = BeautifulSoup(response.text, 'html.parser')
    
    # Find the table containing GameStop Revenue
    table = soup.find_all('table')[1]  # Assuming the table is the second table on the page
    
    # Extract table headers
    headers = [header.text.strip() for header in table.find_all('th')]
    
    # Extract table rows
    rows = []
    for row in table.find_all('tr'):
        rows.append([val.text.strip() for val in row.find_all('td')])
    
    # Create DataFrame
    gme_revenue = pd.DataFrame(rows[1:], columns=headers)
    
    # Clean up the Revenue column (remove commas and dollar signs)
    gme_revenue['Revenue'] = gme_revenue['Revenue'].str.replace(',', '').str.replace('$', '')
    
    # Convert Revenue column to numeric
    gme_revenue['Revenue'] = pd.to_numeric(gme_revenue['Revenue'])
    
    # Display the last five rows of the DataFrame
    print("Last five rows of DataFrame gme_revenue:")
    print(gme_revenue.tail())
else:
    print(f"Failed to download webpage. Status code: {response.status_code}")


ValueError: 1 columns passed, passed data had 2 columns

## Question 5: Plot Tesla Stock Graph


Use the `make_graph` function to graph the Tesla Stock Data, also provide a title for the graph. Note the graph will only show data upto June 2021.


<details><summary>Hint</summary>

```

You just need to invoke the make_graph function with the required parameter to print the graphs.The structure to call the `make_graph` function is `make_graph(tesla_data, tesla_revenue, 'Tesla')`.

```
    
</details>


In [32]:
import requests
from bs4 import BeautifulSoup
import pandas as pd
import matplotlib.pyplot as plt

# Define the URL of the webpage containing the Tesla stock data
url = "https://cf-courses-data.s3.us.cloud-object-storage.appdomain.cloud/IBMDeveloperSkillsNetwork-PY0220EN-SkillsNetwork/labs/project/stock.html"

# Send a GET request to the URL
response = requests.get(url)

# Check if the request was successful (status code 200)
if response.status_code == 200:
    # Parse the HTML using BeautifulSoup with html.parser
    soup = BeautifulSoup(response.text, 'html.parser')
    
    # Find all tables on the page
    tables = soup.find_all('table')
    
    # Iterate through each table to find the one that contains Tesla Stock Data
    tesla_stock_data = None
    for table in tables:
        # Check if the table contains headers mentioning Tesla Stock Data
        if 'Tesla' in table.get_text():
            # Extract table rows
            rows = table.find_all('tr')
            
            # Extract table headers
            headers = [header.text.strip() for header in rows[0].find_all('th')]
            
            # Extract table data and create DataFrame
            data = []
            for row in rows[1:]:
                data.append([cell.text.strip() for cell in row.find_all('td')])
            
            # Create DataFrame
            tesla_stock_data = pd.DataFrame(data, columns=headers)
            break
    
    # If no specific table found, handle the error gracefully
    if tesla_stock_data is None:
        print("Could not find Tesla Stock Data table.")
    else:
        # Clean up the Date column (convert to datetime)
        tesla_stock_data['Date'] = pd.to_datetime(tesla_stock_data['Date'])
        
        # Filter data up to June 2021
        tesla_stock_data = tesla_stock_data[tesla_stock_data['Date'] <= '2021-06-30']
        
        # Convert 'Close' column to numeric
        tesla_stock_data['Close'] = pd.to_numeric(tesla_stock_data['Close'], errors='coerce')
        
        # Plotting the data
        plt.figure(figsize=(10, 6))
        plt.plot(tesla_stock_data['Date'], tesla_stock_data['Close'], marker='o', linestyle='-', color='b', label='Tesla Stock Closing Price')
        plt.title('Tesla Stock Closing Price (Up to June 2021)')
        plt.xlabel('Date')
        plt.ylabel('Closing Price ($)')
        plt.xticks(rotation=45)
        plt.grid(True)
        plt.legend()
        plt.tight_layout()
        plt.show()
else:
    print(f"Failed to download webpage. Status code: {response.status_code}")


Could not find Tesla Stock Data table.


## Question 6: Plot GameStop Stock Graph


Use the `make_graph` function to graph the GameStop Stock Data, also provide a title for the graph. The structure to call the `make_graph` function is `make_graph(gme_data, gme_revenue, 'GameStop')`. Note the graph will only show data upto June 2021.


<details><summary>Hint</summary>

```

You just need to invoke the make_graph function with the required parameter to print the graphs.The structure to call the `make_graph` function is `make_graph(gme_data, gme_revenue, 'GameStop')`

```
    
</details>


In [34]:
import requests
from bs4 import BeautifulSoup
import pandas as pd
import matplotlib.pyplot as plt

# Function to retrieve and parse GameStop Stock Data
def retrieve_gme_stock_data(url):
    response = requests.get(url)
    if response.status_code == 200:
        soup = BeautifulSoup(response.text, 'html.parser')
        # Find the table containing GameStop Stock Data
        table = soup.find('table', {'id': 'gamestop-stock-data'})  # Assuming the table has an id attribute 'gamestop-stock-data'
        
        # Check if table is found
        if table is None:
            print("Could not find GameStop Stock Data table.")
            return None
        
        # Extract table rows
        rows = table.find_all('tr')
        
        # Extract table headers
        headers = [header.text.strip() for header in rows[0].find_all('th')]
        
        # Extract table data and create DataFrame
        data = []
        for row in rows[1:]:
            data.append([cell.text.strip() for cell in row.find_all('td')])
        
        # Create DataFrame
        gme_stock_data = pd.DataFrame(data, columns=headers)
        
        # Clean up the Date column (convert to datetime)
        gme_stock_data['Date'] = pd.to_datetime(gme_stock_data['Date'])
        
        # Filter data up to June 2021
        gme_stock_data = gme_stock_data[gme_stock_data['Date'] <= '2021-06-30']
        
        # Convert 'Close' column to numeric
        gme_stock_data['Close'] = pd.to_numeric(gme_stock_data['Close'], errors='coerce')
        
        return gme_stock_data
    else:
        print(f"Failed to download webpage for GameStop Stock Data. Status code: {response.status_code}")
        return None

# Function to retrieve and parse GameStop Revenue Data
def retrieve_gme_revenue_data(url):
    response = requests.get(url)
    if response.status_code == 200:
        soup = BeautifulSoup(response.text, 'html.parser')
        # Find the table containing GameStop Revenue Data
        table = soup.find('table', {'id': 'gamestop-revenue'})  # Assuming the table has an id attribute 'gamestop-revenue'
        
        # Check if table is found
        if table is None:
            print("Could not find GameStop Revenue Data table.")
            return None
        
        # Extract table rows
        rows = table.find_all('tr')
        
        # Extract table headers
        headers = [header.text.strip() for header in rows[0].find_all('th')]
        
        # Extract table data and create DataFrame
        data = []
        for row in rows[1:]:
            data.append([cell.text.strip() for cell in row.find_all('td')])
        
        # Create DataFrame
        gme_revenue = pd.DataFrame(data, columns=headers)
        
        # Clean up the Revenue column (remove commas and dollar signs)
        gme_revenue['Revenue'] = gme_revenue['Revenue'].str.replace(',', '').str.replace('$', '')
        
        # Convert Revenue column to numeric
        gme_revenue['Revenue'] = pd.to_numeric(gme_revenue['Revenue'], errors='coerce')
        
        return gme_revenue
    else:
        print(f"Failed to download webpage for GameStop Revenue Data. Status code: {response.status_code}")
        return None

# Define the URLs of the webpages containing GameStop stock and revenue data
stock_url = "https://cf-courses-data.s3.us.cloud-object-storage.appdomain.cloud/IBMDeveloperSkillsNetwork-PY0220EN-SkillsNetwork/labs/project/stock.html"
revenue_url = "https://cf-courses-data.s3.us.cloud-object-storage.appdomain.cloud/IBMDeveloperSkillsNetwork-PY0220EN-SkillsNetwork/labs/project/stock.html"

# Retrieve GameStop stock data
gme_stock_data = retrieve_gme_stock_data(stock_url)

# Retrieve GameStop revenue data
gme_revenue = retrieve_gme_revenue_data(revenue_url)

# Function to make graph
def make_graph(gme_data, gme_revenue, title):
    if gme_data is None or gme_revenue is None:
        print("Unable to plot graph. Data retrieval unsuccessful.")
        return
    
    plt.figure(figsize=(12, 6))
    
    # Plot GameStop Stock Data
    plt.subplot(1, 2, 1)
    plt.plot(gme_data['Date'], gme_data['Close'], marker='o', linestyle='-', color='b', label='GameStop Stock Closing Price')
    plt.title('GameStop Stock Data')
    plt.xlabel('Date')
    plt.ylabel('Closing Price ($)')
    plt.xticks(rotation=45)
    plt.grid(True)
    plt.legend()
    
    # Plot GameStop Revenue Data
    plt.subplot(1, 2, 2)
    plt.plot(gme_revenue['Date'], gme_revenue['Revenue'], marker='o', linestyle='-', color='r', label='GameStop Revenue')
    plt.title('GameStop Revenue Data')
    plt.xlabel('Date')
    plt.ylabel('Revenue ($)')
    plt.xticks(rotation=45)
    plt.grid(True)
    plt.legend()
    
    # Adjust layout and show plot
    plt.suptitle(title)
    plt.tight_layout()
    plt.show()

# Call make_graph function to plot GameStop data
make_graph(gme_stock_data, gme_revenue, 'GameStop Stock and Revenue Data (Up to June 2021)')


Could not find GameStop Stock Data table.
Could not find GameStop Revenue Data table.
Unable to plot graph. Data retrieval unsuccessful.


<h2>About the Authors:</h2> 

<a href="https://www.linkedin.com/in/joseph-s-50398b136/">Joseph Santarcangelo</a> has a PhD in Electrical Engineering, his research focused on using machine learning, signal processing, and computer vision to determine how videos impact human cognition. Joseph has been working for IBM since he completed his PhD.

Azim Hirjani


## Change Log

| Date (YYYY-MM-DD) | Version | Changed By    | Change Description        |
| ----------------- | ------- | ------------- | ------------------------- |
| 2022-02-28        | 1.2     | Lakshmi Holla | Changed the URL of GameStop |
| 2020-11-10        | 1.1     | Malika Singla | Deleted the Optional part |
| 2020-08-27        | 1.0     | Malika Singla | Added lab to GitLab       |

<hr>

## <h3 align="center"> © IBM Corporation 2020. All rights reserved. <h3/>

<p>
